### PR TITLE
Add primary term to doc write response

### PR DESCRIPTION
**Pull request by jasontedor**
This commit adds the primary term to the doc write response.

Relates #10708

### CODE REVIEW

File: docs/reference/docs/bulk.asciidoc [hunk 1]
```diff
@@ -154,13 +157,14 @@ The result of this bulk operation is:
...
-            "_seq_no" : 3
+            "_seq_no" : 3,
+            "_primary_term" : 4
          }
       }
    ]
 }
 --------------------------------------------------
-// TESTRESPONSE[s/"took": 30/"took": $body.took/ s/"index_uuid": .../"index_uuid": $body.items.3.update.error.index_uuid/ s/"_seq_no" : 0/"_seq_no" : $body.items.0.index._seq_no/ s/"_seq_no" : 1/"_seq_no" : $body.items.1.delete._seq_no/ s/"_seq_no" : 2/"_seq_no" : $body.items.2.create._seq_no/ s/"_seq_no" : 3/"_seq_no" : $body.items.3.update._seq_no/]
+// TESTRESPONSE[s/"took": 30/"took": $body.took/ s/"index_uuid": .../"index_uuid": $body.items.3.update.error.index_uuid/ s/"_seq_no" : 0/"_seq_no" : $body.items.0.index._seq_no/ s/"_primary_term" : 1/"_primary_term" : $body.items.0.index._primary_term/ s/"_seq_no" : 1/"_seq_no" : $body.items.1.delete._seq_no/ s/"_primary_term" : 2/"_primary_term" : $body.items.1.delete._primary_term/ s/"_seq_no" : 2/"_seq_no" : $body.items.2.create._seq_no/ s/"_primary_term" : 3/"_primary_term" : $body.items.2.create._primary_term/ s/"_seq_no" : 3/"_seq_no" : $body.items.3.update._seq_no/ s/"_primary_term" : 4/"_primary_term" : $body.items.3.update._primary_term/]
```
Comment (nik9000) [inline docs/reference/docs/bulk.asciidoc]: You can stick these onto separate lines if you like.

Comment (jasontedor) [inline docs/reference/docs/bulk.asciidoc]: I pushed 4c6b35cfb9c9db9de962013cf036f8e6f9056a1c.

File: core/src/main/java/org/elasticsearch/index/engine/Engine.java [hunk 2]
```diff
@@ -307,20 +307,22 @@ public Condition newCondition() {
...
             this.failure = failure;
             this.version = version;
             this.seqNo = seqNo;
+            this.primaryTerm = primaryTerm;
         }
 
-        protected Result(Operation.TYPE operationType, long version, long seqNo) {
-            this(operationType, null, version, seqNo);
+        protected Result(Operation.TYPE operationType, long version, long seqNo, long primaryTerm) {
+            this(operationType, null, version, seqNo, primaryTerm);
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/engine/Engine.java]: I'm on sure about this one - the semantics we currently have is that Result represents everything that the engine decides. The primary term is part of the operation - I'm not sure we need this here?

File: docs/reference/docs/bulk.asciidoc [hunk 3]
```diff
@@ -154,13 +157,23 @@ The result of this bulk operation is:
...
-            "_seq_no" : 3
+            "_seq_no" : 3,
+            "_primary_term" : 4
          }
       }
    ]
 }
 --------------------------------------------------
-// TESTRESPONSE[s/"took": 30/"took": $body.took/ s/"index_uuid": .../"index_uuid": $body.items.3.update.error.index_uuid/ s/"_seq_no" : 0/"_seq_no" : $body.items.0.index._seq_no/ s/"_seq_no" : 1/"_seq_no" : $body.items.1.delete._seq_no/ s/"_seq_no" : 2/"_seq_no" : $body.items.2.create._seq_no/ s/"_seq_no" : 3/"_seq_no" : $body.items.3.update._seq_no/]
+// TESTRESPONSE[s/"took": 30/"took": $body.took/]
```
Comment (bleskes) [inline docs/reference/docs/bulk.asciidoc]: good to know this is possible

**Review by bleskes — CHANGES_REQUESTED** [review]
Looks good - I left one comment about the Engine related classes.

File: core/src/main/java/org/elasticsearch/index/engine/Engine.java (hunk 2 above)
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/index/engine/Engine.java]: I think you mean d4bbfcde9713a234d2e82d6ac5dba97ab696adf9?

**Review by bleskes — APPROVED** [review]

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 4]
```diff
@@ -1009,8 +1006,8 @@ private DeleteResult deleteInLucene(Delete delete, DeletionStrategy plan)
         } catch (Exception ex) {
             if (indexWriter.getTragicException() == null) {
                 // there is no tragic event and such it must be a document level failure
-                return new DeleteResult(ex, plan.versionOfDeletion, plan.versionOfDeletion,
-                    plan.currentlyDeleted == false);
+                return new DeleteResult(
+                        ex, plan.versionOfDeletion, plan.seqNoOfDeletion, plan.currentlyDeleted == false);
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: @bleskes By the way, I think this was a bug, maybe there's a test case missing here?

File: docs/reference/docs/bulk.asciidoc (hunk 3 above)
Comment (jasontedor) [inline docs/reference/docs/bulk.asciidoc]: Yes, I didn't know it was until @nik9000 made the suggestion and so I went and read the source to understand the syntax behind it.

**Comment by jasontedor** [comment]
Thanks @nik9000 and @bleskes.

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java (hunk 4 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: @areek I think that the tests in https://github.com/elastic/elasticsearch/pull/23314 don't cover the delete case (but only indexing). Can you look at this?

Comment (areek) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: @bleskes thanks for the ping, I am working on adding tests for the delete case

Comment (areek) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: I added tests for this https://github.com/elastic/elasticsearch/commit/077a6c3ee7ffe0a1b5b4ea7718adf0429409171a

**Pull request merged**